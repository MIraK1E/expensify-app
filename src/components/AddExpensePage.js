import React from 'react'
import { connect } from 'react-redux'
import { startAddExpense } from '../actions/expenses'
import ExpenseForm from './ExpenseForm'

// Avoid function inline by re-write to class base component
class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startAddExpense(expense)
        this.props.history.push('/')
    }
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm 
                    // function excute when call only
                    onSubmit={this.onSubmit}
                    />
                </div>
            </div>
        )
    }
}

// To Avoid function in line that recalculate every time  
// const AddExpensePage = (props) => (
//     <div>
//         <h1>Add Expense</h1>
//         <ExpenseForm 

//             // pass props onSubmit function to child node
//             onSubmit={(expense)=> {

//                 // after mapDispatchToProps props.dispatch(addExpense(expense))
//                 // change to line below and work correctly but This is function inline
//                 props.onSubmit(expense)

//                 // props.htistory come from reactRounter that pass this automatically
//                 props.history.push('/')
//             }}
//         />
//     </div>
// )

// get dispatch function to prop that can be function ready to use 
const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
})

// export to test
export { AddExpensePage }

// connect(mapStateToProp, mapDispatchToProps, ...)
export default connect(undefined, mapDispatchToProps)(AddExpensePage)