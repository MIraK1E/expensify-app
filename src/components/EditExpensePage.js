import React from 'react'
import ExpenseForm from './ExpenseForm'
import { connect } from 'react-redux'
import { startEditExpense, startRemoveExpense } from '../actions/expenses'

class EditExpensePage extends React.Component {
    editExpense = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense)
        this.props.history.push('/dashboard')
    }
    removeExpense = () => {
        this.props.startRemoveExpense({ id: this.props.expense.id })
        this.props.history.push('/dashboard')
    }
    render() {
        return (
            <div>
                <ExpenseForm 
                    expense = {this.props.expense}
                    onSubmit = {this.editExpense}
                />
                <button 
                    onClick={this.removeExpense}>Remove</button>
                </div>
        )
    }
}


// const EditExpensePage = (props) => {
//     return (
//         <div>
//            <ExpenseForm 
//                 expense = {props.expense}
//                 onSubmit = {(expense) => {
//                     props.dispatch(editExpense(props.expense.id, expense))
//                     props.history.push('/')
//                 }}
//            />
//            <button 
//                 onClick={() => { 
//                     props.dispatch(removeExpense({ id: props.expense.id }))
//                     props.history.push('/')
//                 }
//             }>Remove</button>
//         </div>
//     )
// }

const mapStateToProps = (state, props) => {
    return { 
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    }
}

const mapDispatchToProps = (dispatch, props) => {
   return {
       startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
       startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
   }
}

export { EditExpensePage }

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)