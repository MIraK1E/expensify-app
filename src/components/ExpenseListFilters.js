import React from 'react'
import { connect } from 'react-redux'
import { DateRangePicker } from 'react-dates'
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters'
// line above have import an action generator to use 

class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    }
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }
    onFocusedChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }))
    }
    onTextChange = (e) => {
                    // react-redux has send function dispatch to prop then we can dispatch here 
                    // use action generator  
                    // this.props.dispatch(setTextFilter(e.target.value))
                    this.props.setTextFilter(e.target.value)
    }
    onSortChange = (e) => {
        if(e.target.value === 'Date')
        {
            this.props.sortByDate()
        }
        else if(e.target.value === 'Amount')
        {
            this.props.sortByAmount()
        }
    }
    render() {
        return (
            <div>
                <input type="text" value={ this.props.filters.text } onChange={this.onTextChange}/>
                <select 
                    value = {this.props.filters.sortBy}
                    onChange={this.onSortChange}>
                    <option value="Date">Date</option>
                    <option value="Amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate = {this.props.filters.startDate}
                    endDate = {this.props.filters.endDate} 
                    onDatesChange = {this.onDatesChange}
                    focusedInput = {this.state.calendarFocused}
                    onFocusChange = {this.onFocusedChange}
                    showClearDates = {true}
                    numberOfMonths = {1}
                    isOutsideRange = {() => false}
                />
            </div>
        )
    }
}

const mapStateToProps = (state) => ({
        filters: state.filters
})

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
})

export { ExpenseListFilters }

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)