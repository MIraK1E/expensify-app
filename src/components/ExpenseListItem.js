import React from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import moment from 'moment'
import numeral from 'numeral'

// const ExpenseListItem = ({ dispatch, id, description, amount, createdAt }) => (
//     <div>
//          <Link to={`/edit/${id}`}><h3>{ description }</h3></Link>
//         <p>{ amount } - { createdAt }</p>
//         <button onClick={() => actionRemove(dispatch, id)}>Remove</button>
//     </div>
// )

// we can break function and pass argument to excute
// const actionRemove = (dispatch, id) => {
//     dispatch(removeExpense({id}))
// }

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
        <Link to={`/edit/${id}`} className="list-item">
            <div>
                <h3 className="list-item__title">{description}</h3>
                <span className="list-item__sub-title">{moment(createdAt).format('MMMM Do, YYYY')}</span>
            </div>
            <h3 className="list-item__title">
                {numeral(amount / 100).format('$0,0.00')}
            </h3>
        </Link>
)

export default ExpenseListItem