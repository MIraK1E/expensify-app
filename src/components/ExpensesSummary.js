import React from 'react'
import numeral from 'numeral'
import selectExpensesTotal from '../selector/expense-total'
import { connect } from 'react-redux'
import getVisibleExpenses from '../selector/expenses'
import { Link } from 'react-router-dom'

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
    const ExpenseText = expenseCount === 1 ? "expense" : "expenses"
    return (
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">
                    Viewing <span>{expenseCount}</span> {ExpenseText} totalling <span>{numeral(expensesTotal / 100).format('$0,0.00')}</span>
                </h1>
                <div className="page-header__actions">
                    <Link className="button" to="/create">Add Expense</Link>
                </div>
            </div>
        </div>
    )
}

const mapStateToProps = (state) => {
    const visibility = getVisibleExpenses(state.expenses, state.filters)
    return {
        expensesTotal: selectExpensesTotal(visibility),
        expenseCount:  visibility.length
    }
}

export default connect(mapStateToProps)(ExpensesSummary)