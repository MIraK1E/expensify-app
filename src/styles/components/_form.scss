.form {
    display: flex;
    flex-direction: column;
    // mean all children in . form
    > * {
        margin-bottom: $m-size;
    }
}

.form__error {
    margin: 0 0 $m-size 0;
    font-style: italic;
}