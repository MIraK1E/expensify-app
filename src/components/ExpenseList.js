import React from 'react'
// this import it allow us to connect data store
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import getVisibleExpenses from '../selector/expenses'

// props come from ConnectedExpenseList(HOC) that connect to state and send props to this component
// export for test
export const ExpenseList = (props) => (
    <div className="content-container">
    <div>
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>
    </div>
        <div className="list-body">
            {
            props.expenses.length === 0 ? (
                <div>
                    <span className="list-item list-item--message">No expenses</span>
                </div>
            ) : (
                    props.expenses.map((expense) =>
                        <ExpenseListItem
                            key={expense.id}
                            id={expense.id}
                            description={expense.description}
                            amount={expense.amount}
                             createdAt={expense.createdAt}
                        />)
                )
            }
        </div>
    </div>
)

// we can break in to variable like this or just set like comment code
// const mapStateToProps = (state) => {
//     return {
//          //key: the state object that need here
//          expenses: state.expenses,
//          filters: state.filters
//     }
// } 
// export default connect(mapStateToProps)(ExpenseList)
// patern of api connect connect(...function())(component)
const ConnectedExpenseList = connect((state) => {
    return {
        //key: the state object that need here
        //expenses: state.expenses
        //line below has import filter function from selector that need 2 arr to sort and filter then return object back
        expenses: getVisibleExpenses(state.expenses, state.filters)
    }
})(ExpenseList)

export default ConnectedExpenseList