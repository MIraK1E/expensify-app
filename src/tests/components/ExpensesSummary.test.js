import React from 'react'
import { shallow } from 'enzyme'
import { ExpensesSummary } from '../../components/ExpensesSummary'

test('should render a ExpensesSummary with 1 expense data', () => {
    const wrapper = shallow(<ExpensesSummary expensesTotal={100} expenseCount={1} />)
    expect(wrapper).toMatchSnapshot()
})

test('should render a ExpensesSummary with multiple expense data', () => {
    const wrapper = shallow(<ExpensesSummary expensesTotal={5100} expenseCount={2} />)
    expect(wrapper).toMatchSnapshot()
})

