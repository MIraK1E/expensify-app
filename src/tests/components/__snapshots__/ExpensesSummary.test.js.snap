// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a ExpensesSummary with 1 expense data 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        $1.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render a ExpensesSummary with multiple expense data 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        2
      </span>
       
      expenses
       totalling 
      <span>
        $51.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
